### ISSUE TITLE
Memory leak in  OkHttp client  when operation is cancelled

### ISSUE THREAD

**Issue by mhawrylczak** [ktorio/ktor#1177]
### Ktor Version
1.2.1
### Ktor Engine Used(client or server and name)
client OkHttp
### JVM Version, Operating System and Relevant Context
jvm 1.8, linux, osx, 
### Feedback
The OkHttp connections are leaking when operation is cancelled.
```
import io.ktor.application.call
import io.ktor.client.HttpClient
import io.ktor.client.engine.okhttp.OkHttp
import io.ktor.client.request.get
import io.ktor.client.request.url
import io.ktor.http.ContentType
import io.ktor.http.HttpStatusCode
import io.ktor.response.respondTextWriter
import io.ktor.routing.get
import io.ktor.routing.routing
import io.ktor.server.engine.embeddedServer
import io.ktor.server.netty.Netty
import kotlinx.coroutines.delay
import kotlinx.coroutines.runBlocking
import kotlinx.coroutines.withTimeout
import java.util.concurrent.TimeUnit

fun main() = runBlocking {
    val chunk = (1..1024).joinToString(separator = "") { "0" }
    val server = embeddedServer(Netty, 8080) {
        routing {
            get("/") {
                call.respondTextWriter(contentType = ContentType.Text.Plain, status = HttpStatusCode.OK) {
                    repeat(100) {
                        write(chunk)
                        flush()
                        delay(1)
                    }
                    delay(10000)
                }
            }
        }
    }

    server.start(false)
    try {
        val client = HttpClient(OkHttp)
        (1..100000).forEach {
            runCatching {
                withTimeout(100) {
                    client.get<String> { url("http://localhost:8080/") }
                }
            }
            if (it % 100 == 0) {
                val runtime = Runtime.getRuntime()
                runtime.gc()
                val memory = runtime.totalMemory() - runtime.freeMemory()
                println("memory used $memory")
            }
        }
    } finally {
        server.stop(1, 1, TimeUnit.SECONDS)
    }
}
```

**Comment by e5l** [comment]
Hi @mhawrylczak, thanks for the report.
`HttpClient` requires `close()` call. Can you check it with close?

**Comment by mhawrylczak** [comment]
Hi
I have doubt if the example really mirrors the problem.
And I am not sure the ktor is source of the leak.
From time to time i am getting logs like:

> 
> A connection to http://xx.xx.xx.xxx:xxxxx/ was leaked. Did you forget to close a response body? To see where this was allocated, set the OkHttpClient logger level to FINE: Logger.getLogger(OkHttpClient.class.getName()).setLevel(Level.FINE);

For sure the mentioned connections was created by ktor client , but I am not sure if this really is causing  leak.
The problem is visible only sometimes, when service is highly loaded with many timeouts/ cancellations. In such cases class histogram suggests  okhttp leak. 

Would it be possible that this line is a reason ?
https://github.com/ktorio/ktor/blob/master/ktor-client/ktor-client-okhttp/jvm/src/io/ktor/client/engine/okhttp/OkHttpEngine.kt#L76


Documentation for the overloaded version
> public fun invokeOnCompletion(
   onCancelling: Boolean = false,
   invokeImmediately: Boolean = true,
   handler: CompletionHandler): DisposableHandle

suggesst that the handler is not beign invoked if the job is not completed 
> onCancelling ... when `false` then the [handler] is invoked only when it transitions to _completed_ state.

**Comment by dmitrievanthony** [comment]
Hi @mhawrylczak, 

Well, I spent some time trying to reproduce the issue using your example. I used VisualVM to observe memory usage and I saw the following picture: 

<img width="850" alt="Screenshot 2020-01-24 at 10 24 38" src="https://user-images.githubusercontent.com/1028969/73051416-bec64180-3e93-11ea-8b6c-7d5175eb34dd.png">

So, taking this into account I would say that there is no memory leak.

Speaking about the logging:

> A connection to http://xx.xx.xx.xxx:xxxxx/ was leaked. Did you forget to close a response body? To see where this was allocated, set the OkHttpClient logger level to FINE: Logger.getLogger(OkHttpClient.class.getName()).setLevel(Level.FINE);

I'm able to reproduce it using version `1.2.1`, but it works using the current master. So I could conclude that it's been fixed.
